var/spack/repos/builtin/packages/plumed/package.py

```diff
@@ -73,57 +75,46 @@ class Plumed(AutotoolsPackage):
     depends_on('automake', type='build')
     depends_on('libtool', type='build')
 
-    # Dictionary mapping PLUMED versions to the patches it provides
-    # interactively
-    plumed_patches = {
-        '2.3.0': {
-            'amber-14': '1',
-            'gromacs-2016.1': '2',
-            'gromacs-4.5.7': '3',
-            'gromacs-5.0.7': '4',
-            'gromacs-5.1.4': '5',
-            'lammps-6Apr13': '6',
-            'namd-2.8': '7',
-            'namd-2.9': '8',
-            'espresso-5.0.2': '9'
-        },
-        '2.2.4': {
-            'amber-14': '1',
-            'gromacs-4.5.7': '2',
-            'gromacs-4.6.7': '3',
-            'gromacs-5.0.7': '4',
-            'gromacs-5.1.2': '5',
-            'lammps-6Apr13': '6',
-            'namd-2.8': '7',
-            'namd-2.9': '8',
-            'espresso-5.0.2': '9'
-        },
-        '2.2.3': {
-            'amber-14': '1',
-            'gromacs-4.5.7': '2',
-            'gromacs-4.6.7': '3',
-            'gromacs-5.0.7': '4',
-            'gromacs-5.1.2': '5',
-            'lammps-6Apr13': '6',
-            'namd-2.8': '7',
-            'namd-2.9': '8',
-            'espresso-5.0.2': '9'
-        }
-    }
-
     force_autoreconf = True
 
     parallel = False
 
     def apply_patch(self, other):
-        plumed = subprocess.Popen(
-            [join_path(self.spec.prefix.bin, 'plumed'), 'patch', '-p'],
-            stdin=subprocess.PIPE
+
+        # The name of MD engines differ slightly from the ones used in Spack
+        format_strings = collections.defaultdict(
+            lambda: '{0.name}-{0.version}'
         )
-        opts = Plumed.plumed_patches[str(self.version)]
-        search = '{0.name}-{0.version}'.format(other)
-        choice = opts[search] + '\n'
-        plumed.stdin.write(choice)
+        format_strings['espresso'] = 'q{0.name}-{0.version}'
+        format_strings['amber'] = '{0.name}{0.version}'
+
+        get_md = lambda x: format_strings[x.name].format(x)
+
+        # Get available patches
+        plumed = subprocess.Popen([
+            join_path(self.spec.prefix.bin, 'plumed-patch'),
+            '-q',
+            '-l'
+        ], stdout=subprocess.PIPE)
+        available = plumed.stdout.readlines()
+        available = [x.strip() for x in available]
+        plumed.wait()
+
+        # Check that `other` is among the patchable applications
+        if get_md(other) not in available:
+            msg = '{0.name}@{0.version} is not among the MD engine'
+            msg += ' that can be patched by {1.name}@{1.version}.\n'
+            for x in available:
+                msg += x + '\n'
+            raise RuntimeError(msg.format(other, self.spec))
+
+        # Call plumed-patch to patch executables
+        plumed = subprocess.Popen([
```

Comment:
why not use 

```python
plumed_patch = Executable(join_path(self.spec.prefix.bin, 'plumed-patch'))
plumed_patch('-p', '-e', ...)
```

seems more readable and you don't have to wait for it.  same for the other popens too.

Reply:
Sure. Is there any way in which I can obtain something similar to `stdout.readlines()`? Or do I need to manipulate the concatenation of `stdout` afterwards?

Reply:
done